dask/dataframe/multi.py

```diff
@@ -610,6 +624,52 @@ def merge(
         if left_on and right_on:
             warn_dtype_mismatch(left, right, left_on, right_on)
 
+        # Check if we should use a broadcast_join
+        # See note on `broadcast_bias` below.
+        broadcast_bias = 0.5
+        if isinstance(broadcast, float):
+            broadcast_bias = float(broadcast)
+            broadcast = None
+        elif not isinstance(broadcast, bool) and broadcast is not None:
+            # Let's be strict about the `broadcast` type to
+            # avoid arbitrarily casting int to float or bool.
+            raise ValueError(
+                f"Optional `broadcast` argument must be float or bool."
+                f"Type={type(broadcast)} is not supported."
+            )
+        bcast_side = "left" if left.npartitions < right.npartitions else "right"
+        n_small = min(left.npartitions, right.npartitions)
+        n_big = max(left.npartitions, right.npartitions)
+        if (
+            shuffle == "tasks"
+            and how in ("inner", "left", "right")
+            and how != bcast_side
+            and not npartitions
+            and broadcast is not False
+        ):
+            # Note on `broadcast_bias`:
+            # We can expect the broadcast merge to be competitive with
+            # the shuffle merge when the number of partitions in the
+            # smaller collection is less than the logarithm of the number
+            # of partitions in the larger collection.  By default, we add
+            # a small preference for the shuffle-based merge by multiplying
+            # the log result by a 0.5 scaling factor.  We call this factor
+            # the `broadcast_bias`, because a larger number will make Dask
+            # more likely to select the `broadcast_join` code path.  If
+            # the user specifies a floating-point value for the `broadcast`
+            # kwarg, that value will be used as the `broadcast_bias`.
+            if broadcast or (n_small < math.log2(n_big) * broadcast_bias):
```

Comment:
@jakirkham - Pinging you here since you [made a related comment earlier](https://github.com/dask/dask/pull/7143#discussion_r569691566).  I simplified the heuristic and added an option for the user to specify a `broadcast_bias` (as `merge(..., broadcast=<float>)`).  This should allow users (and/or dask_cudf) to tune the heuristic a bit.  Note that it probably makes sense to use a larger bias in dask_cudf - I am trying to be a bit more conservative in `dask.dataframe`.

Reply:
Should that live in `dask.config` then or are you still refining this further?

Reply:
We could check `dask.config` for a `broadcast_bias` default if the user doesn't specify anything. Is that what you have in mind?

Reply:
Yea exactly. Though I don't have a good intuition about how broadly this parameter may apply to usage patterns. So happy to defer to you on whether a config parameter makes sense

Reply:
Yeah - I'm honestly not sure.  I'm assuming the best choice depends on specific characteristics of the data as well as the hardware.  Therefore, it probably does makes sense to make the bias configurable (though I would expect 99% of users to pass the `broadcast` arg if default usage wasn't sufficient).

Reply:
Ok if it is specific to the data then yeah a config variable is probably not needed. If we do find it makes sense in the future, we can always add it later